File: fix_errcheck.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

// Parse errcheck output
const output = fs.readFileSync('/tmp/errcheck_output.txt', 'utf8');
const lines = output.split('\n').filter(l => l.trim());

// Group issues by file
const byFile = {};
for (const line of lines) {
    const tabIdx = line.indexOf('\t');
    if (tabIdx === -1) continue;
    const loc = line.substring(0, tabIdx).trim();
    const expr = line.substring(tabIdx + 1).trim();
    const m = loc.match(/^(.+?):(\d+):(\d+):$/);
    if (!m) continue;
    const file = m[1];
    const lineNum = parseInt(m[2]);
    if (!byFile[file]) byFile[file] = [];
    byFile[file].push({ line: lineNum, expr });
}

let totalFixes = 0;

for (const [filepath, issues] of Object.entries(byFile).sort()) {
    if (!fs.existsSync(filepath)) {
        console.log(`SKIP: ${filepath} not found`);
        continue;
    }
    
    const isTest = filepath.endsWith('_test.go');
    const content = fs.readFileSync(filepath, 'utf8');
    let fileLines = content.split('\n');
    
    // Sort issues by line descending for bottom-up processing
    issues.sort((a, b) => b.line - a.line);
    
    let fixesApplied = 0;
    
    for (const issue of issues) {
        const idx = issue.line - 1;
        if (idx < 0 || idx >= fileLines.length) continue;
        
        const line = fileLines[idx];
        const stripped = line.trim();
        
        if (stripped.startsWith('defer ')) {
            // defer foo(args) -> defer func() { _ = foo(args) }()
            const indent = line.substring(0, line.length - line.trimStart().length);
            const deferExpr = stripped.substring(6); // Remove 'defer '
            fileLines[idx] = `${indent}defer func() { _ = ${deferExpr} }()`;
            fixesApplied++;
        } else if (!isTest) {
            // Non-test: _ = expr
            const indent = line.substring(0, line.length - line.trimStart().length);
            fileLines[idx] = `${indent}_ = ${issue.expr}`;
            fixesApplied++;
        } else {
            // Test file: check if we have 't' in enclosing function
            let hasT = false;
            for (let i = idx - 1; i >= Math.max(0, idx - 60); i--) {
                const fl = fileLines[i];
                if (/\*\s*testing\.T\b/.test(fl) || /\*\s*testing\.T\)/.test(fl)) {
                    hasT = true;
                    break;
                }
                if (fl.trim() === 'func ' && i < idx - 5) break;
                if (/^func\s/.test(fl.trim())) {
                    if (/\*\s*testing\.T/.test(fl)) {
                        hasT = true;
                    }
                    break;
                }
            }
            
            const indent = line.substring(0, line.length - line.trimStart().length);
            if (hasT) {
                fileLines[idx] = `${indent}require.NoError(t, ${issue.expr})`;
            } else {
                fileLines[idx] = `${indent}_ = ${issue.expr}`;
            }
            fixesApplied++;
        }
    }
    
    if (fixesApplied > 0) {
        fs.writeFileSync(filepath, fileLines.join('\n'));
        console.log(`Fixed ${fixesApplied} in ${filepath}`);
        totalFixes += fixesApplied;
    }
}

console.log(`\nTotal: ${totalFixes}`);

File: fix_errcheck.py
#!/usr/bin/env python3
"""
Fix errcheck issues across the codebase.
"""
import re
import os
import sys
from collections import defaultdict

def parse_errcheck_output(output_file):
    """Parse errcheck output into structured data."""
    issues = []
    with open(output_file, 'r') as f:
        for line in f:
            line = line.rstrip('\n')
            # Format: file:line:col:\texpression
            parts = line.split('\t', 1)
            if len(parts) < 2:
                continue
            loc, expr = parts
            loc = loc.strip()
            expr = expr.strip()
            # Parse file:line:col
            m = re.match(r'^(.+?):(\d+):(\d+)$', loc)
            if m:
                issues.append({
                    'file': m.group(1),
                    'line': int(m.group(2)),
                    'col': int(m.group(3)),
                    'expr': expr,
                })
    return issues

def is_test_file(filepath):
    return filepath.endswith('_test.go')

def fix_defer_line(line_content):
    """Fix a defer line that has an unchecked error return.
    defer foo(args) -> defer func() { _ = foo(args) }()
    """
    stripped = line_content.lstrip()
    indent = line_content[:len(line_content) - len(stripped)]
    
    # Match: defer <expr>
    m = re.match(r'^(\s*)defer\s+(.+)$', line_content)
    if m:
        indent = m.group(1)
        expr = m.group(2).rstrip()
        return f"{indent}defer func() {{ _ = {expr} }}()\n"
    return None

def fix_non_defer_line_test(line_content, expr, has_t):
    """Fix a non-defer line in a test file."""
    stripped = line_content.lstrip()
    indent = line_content[:len(line_content) - len(stripped)]
    
    if has_t:
        # Try to use require.NoError(t, expr)
        # But check if line already has a comment
        return f"{indent}require.NoError(t, {expr})\n"
    else:
        return f"{indent}_ = {expr}\n"

def fix_non_defer_line_non_test(line_content, expr):
    """Fix a non-defer line in a non-test file."""
    stripped = line_content.lstrip()
    indent = line_content[:len(line_content) - len(stripped)]
    return f"{indent}_ = {expr}\n"

def has_test_t_param(filepath):
    """Check if the test function has 't *testing.T' parameter."""
    # This is a simplified check - we'll determine per-line context instead
    return True  # Default assumption, refined later

def find_enclosing_function_has_t(lines, target_line):
    """Check if the enclosing function at target_line has a 't' parameter of type *testing.T."""
    # Look backwards from target_line for function definition
    for i in range(target_line - 1, max(0, target_line - 50), -1):
        line = lines[i].strip()
        # Look for func pattern with *testing.T
        if re.search(r'func\s+\w+\s*\([^)]*\*testing\.T\b', line):
            # Check if t is captured
            m = re.search(r'\(\s*(\w+)\s+\*testing\.T', line)
            if m:
                return True
        # Stop at function boundaries (closing brace at col 0 or simple closing)
        if line == '}' and i < target_line - 2:
            # Check if there's a function start before
            for j in range(i - 1, max(0, i - 30), -1):
                if 'func ' in lines[j]:
                    break
            else:
                continue
            break
        # Also check for Test function signatures
        if re.match(r'^func\s+Test', line):
            m = re.search(r'\(\s*t\s+\*testing\.T', line)
            if m:
                return True
    # Check if in a method that takes *testing.T
    for i in range(target_line - 1, max(0, target_line - 50), -1):
        if 'func ' in lines[i]:
            m = re.search(r'\(\s*t\s+\*testing\.T', lines[i])
            if m:
                return True
            m2 = re.search(r'\*testing\.T', lines[i])
            if m2:
                return True
            break
    return False

def get_indent(line):
    """Get the whitespace indentation of a line."""
    return line[:len(line) - len(line.lstrip())]

def process_file(filepath, file_issues):
    """Process a single file and fix all errcheck issues."""
    if not os.path.exists(filepath):
        print(f"  WARNING: {filepath} does not exist")
        return 0
    
    with open(filepath, 'r') as f:
        content = f.read()
    
    lines = content.split('\n')
    # Don't add trailing \n since split removes it
    # Keep as-is and rejoin with \n
    
    test_file = is_test_file(filepath)
    
    # Sort issues by line number descending so we can process from bottom to top
    # This prevents line number shifts from affecting subsequent fixes
    file_issues_sorted = sorted(file_issues, key=lambda x: x['line'], reverse=True)
    
    fixes_applied = 0
    lines_modified = set()
    
    for issue in file_issues_sorted:
        line_idx = issue['line'] - 1  # Convert to 0-indexed
        if line_idx < 0 or line_idx >= len(lines):
            print(f"  WARNING: line {issue['line']} out of range in {filepath}")
            continue
        
        line = lines[line_idx]
        stripped = line.strip()
        
        # Skip already modified lines
        if line_idx in lines_modified:
            continue
        
        # Determine if this is a defer statement
        is_defer = stripped.startswith('defer ')
        
        if is_defer:
            new_line = fix_defer_line(line)
            if new_line:
                lines[line_idx] = new_line.rstrip('\n')
                lines_modified.add(line_idx)
                fixes_applied += 1
        else:
            # Non-defer line
            if test_file:
                has_t = find_enclosing_function_has_t(lines, line_idx)
                new_line = fix_non_defer_line_test(line, issue['expr'], has_t)
                if new_line:
                    lines[line_idx] = new_line.rstrip('\n')
                    lines_modified.add(line_idx)
                    fixes_applied += 1
            else:
                new_line = fix_non_defer_line_non_test(line, issue['expr'])
                if new_line:
                    lines[line_idx] = new_line.rstrip('\n')
                    lines_modified.add(line_idx)
                    fixes_applied += 1
    
    if fixes_applied > 0:
        with open(filepath, 'w') as f:
            f.write('\n'.join(lines))
    
    return fixes_applied

def main():
    output_file = '/tmp/errcheck_output.txt'
    issues = parse_errcheck_output(output_file)
    
    # Group by file
    by_file = defaultdict(list)
    for issue in issues:
        by_file[issue['file']].append(issue)
    
    total_fixes = 0
    for filepath, file_issues in sorted(by_file.items()):
        full_path = filepath
        n = process_file(full_path, file_issues)
        if n > 0:
            print(f"Fixed {n} issues in {filepath}")
            total_fixes += n
    
    print(f"\nTotal fixes applied: {total_fixes}")

if __name__ == '__main__':
    main()